bounds check panic on undo after ctrl-deleting text on last line from MultiLineEntry

### Checklist

- [x] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [x] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

### Description

The text entry widget (either single or multi line) panics when you use ctrl-z to undo removing text. (i did ctrl-backspace on the bottom line and then ctrl-z to undo it):

```
panic: runtime error: slice bounds out of range [63:61]

goroutine 1 [running, locked to thread]:
fyne.io/fyne/v2/widget.(*entryModifyAction).sub(...)
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/widget/entry.go:2012
fyne.io/fyne/v2/widget.(*entryModifyAction).Undo(0xc00277f4a0, {0xc002931d40?, 0xc000049928?})
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/widget/entry.go:1991 +0x279
fyne.io/fyne/v2/widget.(*entryUndoStack).Undo(0x48dfcc?, {0xc002931d40?, 0x44f8bf?})
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/widget/entry.go:2074 +0x4c
fyne.io/fyne/v2/widget.(*Entry).Undo(0xc000156008)
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/widget/entry.go:706 +0x2e
fyne.io/fyne/v2/widget.(*Entry).registerShortcut.func1({0xc000156048?, 0xe34c80?})
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/widget/entry.go:1027 +0x17
fyne.io/fyne/v2.(*ShortcutHandler).TypedShortcut(0xc000156048, {0xe1b8e0, 0x1b771c0})
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/shortcut.go:21 +0x78
fyne.io/fyne/v2/widget.(*Entry).TypedShortcut(0x1324320?, {0xe1b8e0?, 0x1b771c0?})
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/widget/entry.go:874 +0x25
fyne.io/fyne/v2/internal/driver/glfw.(*window).triggersShortcut(0xc00013c000, {0xc00297a2e0?, 0xc00297a2e0?}, {0xc00297a2e4?, 0xc000049a28?}, 0x2)
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/internal/driver/glfw/window.go:854 +0x58c
fyne.io/fyne/v2/internal/driver/glfw.(*window).processKeyPressed(0xc00013c000, {0xc00297a2e0, 0x1}, {0xc00297a2e4, 0x1}, 0x34, 0x1, 0x2)
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/internal/driver/glfw/window.go:722 +0x49e
fyne.io/fyne/v2/internal/driver/glfw.(*window).keyPressed(0xc00013c000, 0xc0001b2480?, 0x5a, 0x34, 0x1, 0x2)
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/internal/driver/glfw/window_desktop.go:632 +0x165
github.com/go-gl/glfw/v3.3/glfw.goKeyCB(0xc000002380?, 0x5a, 0x34, 0x1, 0x2)
	/home/mleku/pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20240506104042-037f3cc74f2a/input.go:358 +0x5d
github.com/go-gl/glfw/v3.3/glfw._Cfunc_glfwPollEvents()
	_cgo_gotypes.go:1603 +0x3a
github.com/go-gl/glfw/v3.3/glfw.PollEvents()
	/home/mleku/pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20240506104042-037f3cc74f2a/window.go:931 +0x13
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).pollEvents(...)
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/internal/driver/glfw/loop_desktop.go:22
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).runGL(0xc0002f4730?)
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/internal/driver/glfw/loop.go:140 +0x1aa
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).Run(0xc0002ef550)
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/internal/driver/glfw/driver.go:162 +0x72
fyne.io/fyne/v2/app.(*fyneApp).Run(0xc0002ef600)
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/app/app.go:77 +0x102
fyne.io/fyne/v2/internal/driver/glfw.(*window).ShowAndRun(0xc00013c000)
	/home/mleku/pkg/mod/fyne.io/fyne/v2@v2.6.0/internal/driver/glfw/window.go:222 +0x64
main.main()
	/home/mleku/src/realy.lol/cmd/bunk/main.go:52 +0x32a
```

this is the code which lacks a bounds check:

```go
// Deletes Text
func (i *entryModifyAction) sub(s string) string {
	runes := []rune(s)
	return string(runes[:i.Position]) + string(runes[i.Position+len(i.Text):])
}
```

in this case it looks like the position has gone beyond the end of the slice and needs to be grown first or the end position clamped to the end

### How to reproduce

1. add text to a multi-line entry
2. ctrl-backspace to delete one of the lines
3. ctrl-z to undo
4. profit

### Screenshots

_No response_

### Example code

there isn't code for this, it's just an entry being used

### Fyne version

v2.6.0

### Go compiler version

1.24.2

### Operating system and version

zorin OS 17.2

### Additional Information

_No response_

--- a/widget/entry.go
+++ b/widget/entry.go
@@ -814,7 +814,13 @@ func (e *Entry) deleteWord(right bool) {
 		end += b.begin
 	}
 
-	provider.deleteFromTo(start, end)
+	erased := provider.deleteFromTo(start, end)
+	e.undoStack.MergeOrAdd(&entryModifyAction{
+		Delete:   true,
+		Position: start,
+		Text:     erased,
+	})
+
 	if !right {
 		e.CursorColumn = cursorCol - (end - start)
 	}

--- a/widget/entry_test.go
+++ b/widget/entry_test.go
@@ -2063,6 +2063,26 @@ func TestEntry_UndoRedo_Delete(t *testing.T) {
 	assert.Equal(t, "àbf", entry.Text)
 }
 
+func TestEntry_UndoRedo_DeleteWord(t *testing.T) {
+	entry := widget.NewMultiLineEntry()
+
+	for _, r := range "Line 1\nline 2" {
+		entry.TypedRune(r)
+	}
+	assert.Equal(t, "Line 1\nline 2", entry.Text)
+
+	moveWordModifier := fyne.KeyModifierShortcutDefault
+	if runtime.GOOS == "darwin" {
+		moveWordModifier = fyne.KeyModifierAlt
+	}
+
+	entry.TypedShortcut(&desktop.CustomShortcut{KeyName: fyne.KeyBackspace, Modifier: moveWordModifier})
+	assert.Equal(t, "Line 1\n", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutUndo{})
+	assert.Equal(t, "Line 1\nline 2", entry.Text)
+}
+
 func TestEntry_UndoRedo_Replace(t *testing.T) {
 	entry := widget.NewEntry()
 

